File: src/gui/src/app/classes/events/job/job-progress-event.ts
import { EventLogLevel } from '@app/interfaces/events';
import { formatBytes } from '@app/utils/utils';
import { ListEventsResponse } from '@gen/es/fme/v1/fme_service_pb';
import { JobProgressEvent as ProtoJobProgressEvent } from '@gen/es/fme/v1/job_pb';

export class JobProgressEvent {
    logLevel = EventLogLevel.Info;

    constructor(public id: string, public name: string, public bytesTransferred: number, public totalBytes: number) {
    }

    get logMessage(): string {
        return `Job progress for ${this.name}: ${formatBytes(this.bytesTransferred)}/${formatBytes(this.totalBytes)} transferred`;
    }

    static fromProtobuf(event: ListEventsResponse): JobProgressEvent {
        const evt = event.event.value as ProtoJobProgressEvent;

        return new JobProgressEvent(
            evt.id,
            evt.name,
            Number(evt.bytesTransferred),
            Number(evt.totalBytes),
        );
    }
}

File: src/cli/types/eventtypes/job_progress_event.go
package eventtypes

import (
	"fmt"

	"github.com/awslabs/filemoverexpress/logger"
	"github.com/awslabs/filemoverexpress/types/pbtypes/fme/v1"
	"github.com/awslabs/filemoverexpress/utils"
)

type JobProgressEvent struct {
	Id               string
	Name             string
	BytesTransferred int64
	TotalBytes       int64
}

func (jpe *JobProgressEvent) String() string {
	return fmt.Sprintf("Job progress for %s: %s/%s transferred", jpe.Name, utils.FormatBytes(jpe.BytesTransferred),
		utils.FormatBytes(jpe.TotalBytes))
}

func (*JobProgressEvent) Type() MessageFlags {
	return JobProgressEventType
}

func (*JobProgressEvent) Priority() logger.LogLevel {
	return JobProgressEventPriority
}

func (jpe *JobProgressEvent) ToProtobuf() (fmev1.PbEvent, fmev1.EventType) {
	pbEvent := &fmev1.JobProgressEvent{
		Id:               jpe.Id,
		Name:             jpe.Name,
		BytesTransferred: jpe.BytesTransferred,
		TotalBytes:       jpe.TotalBytes,
	}
	msgEvent := fmev1.ListEventsResponse_JobProgressEvent{JobProgressEvent: pbEvent}
	return &msgEvent, fmev1.EventType_EVENT_TYPE_JOB_PROGRESS_EVENT_TYPE
}
